README.md

# react-native-scrypt

Non blocking and fast scrypt implementation for React Native.

## Introduction

[scrypt](http://www.tarsnap.com/scrypt.html) is a password-based key derivation function designed to make it costly to perform hardware attacks on the derived keys. While there exist scrypt implementations written in javascript they are extremely slow and impractical for use in mobile apps.

This plugin is for use with React Native and allows your application to use scrypt on iOS/Android devices using native C code, achieving orders of magnitude faster calculation. It is based on [libscrypt](https://github.com/technion/libscrypt).

## Getting started

`$ npm install react-native-scrypt --save`

### Mostly automatic installation

`$ react-native link react-native-scrypt`

### Manual installation


#### iOS

1. In XCode, in the project navigator, right click `Libraries` ➜ `Add Files to [your project's name]`
2. Go to `node_modules` ➜ `react-native-scrypt` and add `RNScrypt.xcodeproj`
3. In XCode, in the project navigator, select your project. Add `libRNScrypt.a` to your project's `Build Phases` ➜ `Link Binary With Libraries`
4. Run your project (`Cmd+R`)<

#### Android

1. Open up `android/app/src/main/java/[...]/MainActivity.java`
  - Add `import com.crypho.scrypt.RNScryptPackage;` to the imports at the top of the file
  - Add `new RNScryptPackage()` to the list returned by the `getPackages()` method
2. Append the following lines to `android/settings.gradle`:
  	```
  	include ':react-native-scrypt'
  	project(':react-native-scrypt').projectDir = new File(rootProject.projectDir, 	'../node_modules/react-native-scrypt/android')
  	```
3. Insert the following lines inside the dependencies block in `android/app/build.gradle`:
  	```
      compile project(':react-native-scrypt')
  	```

## Usage
```javascript
import scrypt from 'react-native-scrypt';

// With 'legacy' encoding (default): passwd must be a string, salt must be an array of bytes integers
// With 'hex' encoding: passwd and salt must be string encoded in hexadecimal
// With 'base64' encoding: passwd and salt must be string encoded in base64
// With 'buffer' encoding: passwd and salt must be Buffers (in the sense of [`buffer`](https://github.com/feross/buffer/) package)
// see example/App.js


const result = await scrypt(passwd, salt[, N=16384, r=8, p=1, dkLen=64, encoding='legacy'])
```

## LICENSE

    The MIT License

    Copyright (c) 2017 Crypho AS.

    Permission is hereby granted, free of charge, to any person obtaining a copy
    of this software and associated documentation files (the "Software"), to deal
    in the Software without restriction, including without limitation the rights
    to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
    copies of the Software, and to permit persons to whom the Software is
    furnished to do so, subject to the following conditions:

    The above copyright notice and this permission notice shall be included in
    all copies or substantial portions of the Software.

    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
    IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
    AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
    OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
    THE SOFTWARE.

    libscrypt is Copyright (c) 2013, Joshua Small under the BSD license. See src/libscrypt/LICENSE


CHANGELOG
Changelog
=========

1.2.1 - 2021-3-19
-----------------

- Updated typings.
  [tex0l]
- Changed the version of the `react-native` peerDependency to `*` to accommodate npm@7's new behaviour.
  [tex0l]

1.2.0 - 2020-10-23
------------------

- Unified native I/O to hex strings ;
  [tex0l]
- Added other encoding methods : 'legacy' which behaves like v1.1.2, 'hex', 'base64' and 'buffer'
  [tex0l]
- Update example with tests for a range of test vectors
  [tex0l]

1.1.2 - 2019-12-20
------------------

- Load Android Gradle Plugin conditionally.
  [SaeedZhiany]

1.1.1 - 2019-11-18
------------------

- Update android config.
  [danh, mo22]

- Update example.
  [danh]

- Update iOS config.
  [maddijoyce]


example/App.js
/**
 * Sample React Native App
 * https://github.com/facebook/react-native
 *
 * @format
 * @flow
 */

import React, {Component} from 'react'
import {ActivityIndicator, Button, StyleSheet, Text, View} from 'react-native'
import scrypt from 'react-native-scrypt'
import sjcl from './sjcl'
import {randomBytes} from 'react-native-randombytes'
import {assert} from 'chai'
import {Buffer} from 'buffer'
import testVectors from './test_vectors'

function seeded() {
  return new Promise(resolve => {
    if (sjcl.random.isReady()) {
      resolve()
    } else {
      sjcl.random.addEventListener('seeded', resolve)
    }
  })
}

export default class App extends Component {
  constructor(...args) {
    super(...args)
    this.state = {
      sjcl: '...',
      libscrypt: '...',
      ready: false,
      tests: 0,
      passedTests: 0,
    }
  }

  async test(name, callback) {
    console.log(`Executing ${name}...`)
    const start = Date.now()
    try {
      await callback()
      console.log(`Test ${name} passed in ${Date.now() - start}ms`)
      this.setState({tests: this.state.tests + 1, passedTests: this.state.passedTests + 1})
    } catch (error) {
      console.warn(`Test ${name} failed in ${Date.now() - start}ms with error`, error)
      this.setState({tests: this.state.tests + 1})
    }
  }

  async componentDidMount() {
    await seeded()
    await this.generateRandomTestVector()
    await this.runExample()
  }

  async generateRandomTestVector() {
    // add a random test vector with non-ASCII characters
    console.log('generating random test vector')
    const testVector = {
      password: await randomBytes(64),
      salt: await randomBytes(64),
      N: 16384,
      r: 8,
      p: 1,
      dkLen: 64,
      skipInLegacy: true,
    }
    testVector.expected = Buffer.from(
      sjcl.codec.hex.fromBits(
        sjcl.misc.scrypt(
          sjcl.codec.hex.toBits(testVector.password.toString('hex')),
          sjcl.codec.hex.toBits(testVector.salt.toString('hex')),
          testVector.N,
          testVector.r,
          testVector.p,
          testVector.dkLen * 8,
        ),
      ),
      'hex',
    )
    testVectors.push(testVector)
    console.log('random test vector generated')
  }

  async runExample() {
    console.log('running quick example', testVectors.length)
    await this.sjclScrypt(testVectors[21])
    console.log('control sjcl scrypt done')
    await this.libScrypt(testVectors[21])
    console.log('native scrypt done')
    this.setState({ready: true})
  }

  async startTests() {
    if (this.state.ready === false) {
      return
    }
    this.setState({ready: false, tests: 0, passedTests: 0})
    console.log('starting tests...')

    for (let i = 0; i < testVectors.length; i++) {
      const {password, salt, N, r, p, dkLen, expected, skipInLegacy = false} = testVectors[i]
      if (!skipInLegacy) {
        await this.test(`Test vector ${i + 1} with encoding legacy`, async () => {
          const encodedPassword = password.toString('utf-8')
          const encodedSalt = sjcl.codec.utf8String.toBits(salt.toString('utf-8'))
          const actualResult = await scrypt(encodedPassword, encodedSalt, N, r, p, dkLen)
          assert.strictEqual(actualResult, expected.toString('hex'))
        })
      }
    }

    for (let i = 0; i < testVectors.length; i++) {
      const {password, salt, N, r, p, dkLen, expected} = testVectors[i]
      await this.test(`Test vector ${i + 1} with encoding base64`, async () => {
        const encodedPassword = password.toString('base64')
        const encodedSalt = salt.toString('base64')
        const actualResult = await scrypt(encodedPassword, encodedSalt, N, r, p, dkLen, 'base64')
        assert.strictEqual(actualResult, expected.toString('base64'))
      })
    }

    for (let i = 0; i < testVectors.length; i++) {
      const {password, salt, N, r, p, dkLen, expected} = testVectors[i]
      await this.test(`Test vector ${i + 1} with encoding hex`, async () => {
        const encodedPassword = password.toString('hex')
        const encodedSalt = salt.toString('hex')
        const actualResult = await scrypt(encodedPassword, encodedSalt, N, r, p, dkLen, 'hex')
        assert.strictEqual(actualResult, expected.toString('hex'))
      })
    }

    for (let i = 0; i < testVectors.length; i++) {
      const {password, salt, N, r, p, dkLen, expected} = testVectors[i]
      await this.test(`Test vector ${i + 1} with encoding buffer`, async () => {
        const encodedPassword = password
        const encodedSalt = salt
        const actualResult = await scrypt(encodedPassword, encodedSalt, N, r, p, dkLen, 'buffer')
        assert.isTrue(actualResult.equals(expected))
      })
    }

    if (this.state.tests === this.state.passedTests) {
      console.log(`All ${this.state.tests} passed.`)
    } else {
      console.error(
        `${this.state.passedTests} / ${this.state.tests} passed, ${this.state.tests - this.state.passedTests} failed.`,
      )
    }
    this.setState({ready: true})
  }

  async sjclScrypt(testVector) {
    const password = sjcl.codec.hex.toBits(testVector.password.toString('hex'))
    const salt = sjcl.codec.hex.toBits(testVector.salt.toString('hex'))
    const N = testVector.N
    const r = testVector.r
    const p = testVector.p
    const dkLen = testVector.dkLen * 8

    const start = Date.now()
    const sjclScrypt = sjcl.misc.scrypt(password, salt, N, r, p, dkLen)
    const elapsed = Date.now() - start

    this.setState({sjcl: `${sjcl.codec.hex.fromBits(sjclScrypt)} ${elapsed}ms`})
  }

  async libScrypt(testVector) {
    const password = testVector.password.toString('utf-8')
    const salt = sjcl.codec.hex.toBits(testVector.salt.toString('hex'))
    const N = testVector.N
    const r = testVector.r
    const p = testVector.p
    const dkLen = testVector.dkLen

    const start = Date.now()
    const libScrypt = await scrypt(password, salt, N, r, p, dkLen)
    const elapsed = Date.now() - start

    this.setState({libscrypt: `${libScrypt} ${elapsed}ms`})
    return libScrypt
  }

  render() {
    return (
      <View style={styles.container}>
        <Text style={styles.text}>{`sjcl: ${this.state.sjcl}`}</Text>
        <Text style={styles.text}>{`react-native-scrypt: ${this.state.libscrypt}`}</Text>
        <Button onPress={() => this.startTests()} title="Start tests" disabled={!this.state.ready} />
        <ActivityIndicator animating={!this.state.ready} />
        <Text style={styles.text}>{`${this.state.passedTests} / ${this.state.tests} passed`}</Text>
        <Text style={styles.text}>{`${this.state.tests - this.state.passedTests} / ${this.state.tests} failed`}</Text>
      </View>
    )
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: '#F5FCFF',
  },
  text: {
    fontSize: 20,
    textAlign: 'center',
    margin: 10,
  },
})
